src/connector/src/sink/dynamodb.rs

```diff
@@ -0,0 +1,329 @@
+                .await?
+                .into_log_sinker(usize::MAX),
+        )
+    }
+}
+
+impl TryFrom<SinkParam> for DynamoDbSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        let schema = param.schema();
+        let config = DynamoDbConfig::from_hashmap(param.properties)?;
+
+        Ok(Self {
+            config,
+            schema,
+            pk_indices: param.downstream_pk,
+        })
+    }
+}
+
+struct DynamoDbPayloadWriter {
+    insert_items: HashMap<String, AttributeValue>,
+    delete_items: HashMap<String, AttributeValue>,
+    client: Client,
+    table: String,
+}
+
+impl DynamoDbPayloadWriter {
+    fn write_one_insert(&mut self, key: String, value: AttributeValue) {
+        self.insert_items.insert(key, value);
+    }
+
+    fn write_one_delete(&mut self, key: String, value: AttributeValue) {
+        self.delete_items.insert(key, value);
+    }
+
+    async fn write_chunk(&mut self) -> Result<()> {
+        if !self.insert_items.is_empty() {
+            let new_items = std::mem::take(&mut self.insert_items);
+            self.client
+                .put_item()
+                .table_name(self.table.clone())
+                .set_item(Some(new_items))
+                .send()
+                .await
+                .map_err(|e| {
+                    SinkError::DynamoDb(anyhow!(e).context("failed to put item to DynamoDB sink"))
+                })
+                .map(|_| ())?
+        }
+        if !self.delete_items.is_empty() {
+            let new_items = std::mem::take(&mut self.delete_items);
+            self.client
+                .delete_item()
+                .table_name(self.table.clone())
+                .set_key(Some(new_items))
+                .send()
+                .await
+                .map_err(|e| {
+                    SinkError::DynamoDb(
+                        anyhow!(e).context("failed to delete item from DynamoDB sink"),
+                    )
+                })
+                .map(|_| ())?
+        }
+
+        Ok(())
+    }
+}
+
+pub struct DynamoDbSinkWriter {
+    payload_writer: DynamoDbPayloadWriter,
+    formatter: DynamoDbFormatter,
+}
+
+impl DynamoDbSinkWriter {
+    pub async fn new(config: DynamoDbConfig, schema: Schema) -> Result<Self> {
+        let client = config.common.build_client().await?;
+
+        let payload_writer = DynamoDbPayloadWriter {
+            insert_items: HashMap::new(),
+            delete_items: HashMap::new(),
+            client,
+            table: config.common.table,
+        };
+
+        Ok(Self {
+            payload_writer,
+            formatter: DynamoDbFormatter { schema },
+        })
+    }
+
+    async fn write_chunk_inner(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            match op {
+                Op::Insert | Op::UpdateInsert => {
```

Comment:
Have you ever tested an `update` event? 

In `self.payload_writer.write_chunk()`, you inserted all the events before delete, which effectively means the `UpdateInsert` will happen `UpdateDelete`. IIUC, the row will be deleted after that rather than being updated.

Here, I prefer to handle `UpdateInsert` and `UpdateDelete` as a whole i.e. convert them into a single `put KV` into DynamoDB.

But keep in mind that this won't really solve the problem. Please carefully consider both of the following case:

- An `Insert` comes after `Delete` on the same key
- An `Delete` comes after `Insert` on the same key

Perhaps the current implementation of `self.payload_writer.write_chunk()` won't work.